### ISSUE TITLE
Migrate the Jupyter Notebooks to Python 3

### ISSUE THREAD

**Issue by BenKaehler** [BenKaehler/short-read-tax-assignment#1]

**Comment by nbokulich** [comment]
One small but potential issue in migrating the taxonomy evaluation notebooks to python 3: the taxonomy assignment module (and possibly others) currently rely on QIIME-1's `assign_taxonomy.py`, which is python 2. Will this not cause a problem?

**Comment by BenKaehler** [comment]
The good news is that `assign_taxonomy.py` is only called via the command line. (I think, please let me know if that isn't the case.) Only yesterday I faced exactly the same problem [here](https://github.com/BenKaehler/q2-feature-classifier-tuning/blob/master/nbk/iceland-rdp.ipynb). I thought there must be a better way, because the way I did it is a bit ugly, but Google has not yet presented me with a better one.

So there are a few options.

Option 1:
1. [Create a Python 2 conda env containing QIIME](http://conda.pydata.org/docs/py2or3.html#create-python-2-or-3-environments)
2. Create a Python 3 conda env containing QIIME 2
3. Run the notebooks from the Python 3 env
4. Whenever you launch `assign_taxonomy.py` make sure you activate the Python 2 env first.

Alternatively, write a small script that wraps the Python 2 `assign_taxonomy.py` script, add it to the repo, and make sure you add its location to your path. The script should simply activate the Python 2 env then pass all its arguments to the real `assign_taxonomy.py`.

This will add a conda dependency to the setup and complicate installation, which is not ideal.

Option 2 is to call RDP Classifier directly and cut out the Python 2 dependency.

Option 3 is to port the `burrito-filling` [wrapper](https://github.com/biocore/burrito-fillings/blob/master/bfillings/rdp_classifier.py) to Python 3.

I would be tempted to try Option 3. It would be slightly more effort than Option 1, but much more usable if anyone ever wants to run the tool again. It looks like [burrito](https://github.com/biocore/burrito) already supports Python 3, so it could be as easy as porting two files (the `burrito-filling` and its unit test) to Python 3.

What do you think?

**Comment by nbokulich** [comment]
I think option 2 or 3, I agree adding a conda dependency would not be ideal. (I will note, though, that once we have the pre-computed classifier results this is no longer a dependency per-se, though ideally we want to preserve a pipeline that others can replicate)

However, the problem goes beyond RDP classifier. It _might_ be nice to add other qiime-1 classifiers, especially the UCLUST classifier, which had performed well previously. I believe @gregcaporaso originally wrote the UCLUST classifier as a stand-alone classifier before wrapping in `assign_taxonomy.py`, so it may be possible to re-write for python-3, but then there is a licensing issue (UCLUST is only free to use in QIIME... does that include QIIME-2, @gregcaporaso ?). I'd still consider RDP the current gold standard, so we could just quietly drop the others and focus on comparing RDP to a selection of new classifiers.

What do you think?

**Comment by BenKaehler** [comment]
It looks to me like [burrito](https://github.com/biocore/burrito) could be used stand-alone to provide RDP, UCLUST, and maybe some other classifiers without too much trouble, but I haven't looked that hard yet. That is if we update the necessary burrito fillings to be Python 3 compliant (Option 3).

For licensing, I don't imagine it would be a problem for us to provide code that uses UCLUST. I would say the onus would be on the user to make sure that her or his installation of UCLUST was appropriately licensed, and, as you say, they only need it if they want to reproduce results that we will distribute with the system.

**Comment by nbokulich** [comment]
Looks like burrito requires an older version of scikit-bio, which is incompatible with other dependencies we have.

We must find an easier way, if the point of the evaluation framework is ease of use and accessibility. Given our current conundrum, I anticipate a high likelihood that others who want to test a new classifier with the evaluation framework may have similar conflicts.

With that in mind, I think that "option 1" from above may be the best choice. It is awkward, but it allows us (and others) to just plug in classifier X in whatever environment is required. If we do go this route, I think it's also worth asking whether it is worth the trouble, then, to migrate the notebooks to python-3. Will this enhance usability/function?

What do you think?

**Comment by nbokulich** [comment]
"Option 1" (call different conda envs to run `qiime1` commands) does not appear to work. As far as I can tell, calling `source activate x` inside Jupyter notebooks does not change the environment. The environment remains as whatever Jupyter was launched from.

However, I have been successful at running these `python3` notebooks if I launch from a `python2` environment that has `qiime1` installed. In the short term, this solves our need, and when it comes to testing `q2-feature-classifier`, I will just need to re-run the notebooks in a `qiime2` environment. 

This should not be a problem for us, as the notebooks that actually compare methods do not require any underlying software and just call pre-computed results. But this could cause minor headaches, e.g., if downstream users decide they want to test classifiers that have conflicting dependencies. This may just be an issue that we document and leave it at that... but if you think there is a fix to make this more straightforward in the long run, we may want to implement.

What do you think?

**Comment by BenKaehler** [comment]
Yes, each time you call a bash command from the notebook it will open a new shell, so you have to activate the new conda environment in the same command in which you invoke the classifier.

A working example is [here](https://github.com/BenKaehler/q2-feature-classifier-tuning/blob/master/nbk/iceland-rdp.ipynb).

An alternative is to make a wrapper, possibly in the same directory as the notebook, possibly called `assign_taxonomy.sh`, along the lines of

```
#!/bin/bash
source activate qiime1
assign_taxonomy.py "$@"
```

and then call `assign_taxonomy.sh` from the notebook.

I am a bit surprised that a Python 3 notebook works inside a Python 2 conda env. I guess there is no python code in it that violates the old specifications.

I'm not sure about switching conda environments to execute different notebooks. It somehow feels like more of a kludge to me than writing scripts to wrap classifiers with conflicting requirements.

**Comment by nbokulich** [comment]
Ah, now I understand. I can work from these examples. Thanks!
